### PR TITLE
FirstCommit

### PR DESCRIPTION
**Pull request by zaini63033**

### CODE REVIEW

File: utils/fetchData.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    {
+        const attributes = Object.keys(query);
+        if(attributes.length > 1)
+        {
+            URL += `?`+ attributes[0] + `=${query[attributes[0]]}&` + attributes[1] + `=${query[attributes[1]]}`;
+        }
+        else if(attributes.length > 0)
+        {
+            URL += `?`+ attributes[0] + `=${query[attributes[0]]}`;
+        }
```
Comment (salmanR107) [inline utils/fetchData.js]: improve this condition. What if i have a third query param?

File: api/fetch-pokemon.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    const data = await fetchData({URL :`https://pokeapi.co/api/v2/pokemon`, query : {limit, offset}});
+    
+    const pokemonDetails = [];
+
+    let id = offset;
+
+    for (const pokemon of data.results) {
+      pokemonDetails.push({id, name: pokemon.name, image : `https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/${id + 1}.png`});
+      ++id;
+    }
```
Comment (salmanR107) [inline api/fetch-pokemon.js]: use the spread operator instead

File: app/components/home/home.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+    const newPokemon = await fetchPokemonData({limit, offset});
+    setPokemon((prevPokemon) => [...prevPokemon, ...newPokemon]);
+
+    setOffset(offset + limit);
+
+    if (offset >= 100000) {
+      setHasMore(false);
+    }
+
```
Comment (salmanR107) [inline app/components/home/home.js]: What is this?

File: app/components/home/home.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+      <h1>Pokémon List</h1>
+      <InfiniteScroll pageStart={0} loadMore={loadMore} hasMore={hasMore}>
+        <div className= {styles["pokemon-container"]}>
+          {pokemon.map((p) => (
+            <div key={p.id} className= {styles["pokemon-item"]}>
+              <Image
+                src={p.image}
+                alt={p.name}
+                width={100}
+                height={100}
```
Comment (salmanR107) [inline app/components/home/home.js]: Use 150

**Review by salmanR107 — CHANGES_REQUESTED** [review]

File: api/fetch-pokemon.js [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+export const fetchPokemonData = async ({ limit = 10, offset = 0 } = {}) => {
+  try {
+    const data = await fetchData({URL :`https://pokeapi.co/api/v2/pokemon`, query : {limit, offset}});
+    
+    const pokemonDetails = [];
+
+    let id = offset;
+
+    for (const pokemon of data.results) {
+      pokemonDetails.push({id, name: pokemon.name, image : `https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/${id + 1}.png`});
```
Comment (salmanR107) [inline api/fetch-pokemon.js]: save backImage too

File: app/components/home/home.js [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+  const [offset, setOffset] = useState(0);
+  const map = {};
+
+  const loadMore = async () => {
+
+    if(offset in map)
+      return
+
+    map[offset] = true;
+
```
Comment (salmanR107) [inline app/components/home/home.js]: Set initialLoad to true
Get the first page's results' from the server side and pass it to HomePage via props.
use the isLoading true/false in the function

File: utils/fetchData.js (hunk 1 above)
Comment (salmanR107) [reply utils/fetchData.js]: let postfix = ""
loop the attributes
If postfix length is 0, add a "?" else "&"
Add the attribute and its value

**Review by salmanR107 — CHANGES_REQUESTED** [review]

**Review by salmanR107 — APPROVED** [review]

File: api/fetch-pokemons.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+        front_image: `${image}${id + 1}.png`,
+        back_image: `${image}/back/${id + 1}.png`,
+      });
+      ++id;
+    }
+
+    return pokemonDetails;
+  } catch (error) {
+    console.error(error);
+    return [];
```
Comment (salmanR107) [inline api/fetch-pokemons.js]: throw error

File: api/fetch-pokemon-data.jsx [hunk 8]
```diff
@@ -0,0 +1,14 @@
+import { fetchData } from '../utils/fetch-data';
+
+export const fetchPokemonData = async ({ name = '' } = {}) => {
+  try {
+    const data = await fetchData({
+      URL: `https://pokeapi.co/api/v2/pokemon/${name}`,
+      query: {},
```
Comment (salmanR107) [inline api/fetch-pokemon-data.jsx]: No need to pass this

File: app/components/home/home.js [hunk 9]
```diff
@@ -2,8 +2,11 @@
 import { useState } from 'react';
 import InfiniteScroll from 'react-infinite-scroller';
 import Image from 'next/image';
-import { fetchPokemonData } from '../../../api/fetch-pokemon';
+import { fetchPokemons } from '../../../api/fetch-pokemons';
 import styles from './home.module.css';
+import {} from 'next/navigation';
+import Link from 'next/link';
+// import { useRouter } from 'next/navigation';
```
Comment (salmanR107) [inline app/components/home/home.js]: Remove

File: app/components/home/home.js [hunk 10]
```diff
@@ -2,8 +2,11 @@
 import { useState } from 'react';
 import InfiniteScroll from 'react-infinite-scroller';
 import Image from 'next/image';
-import { fetchPokemonData } from '../../../api/fetch-pokemon';
+import { fetchPokemons } from '../../../api/fetch-pokemons';
 import styles from './home.module.css';
+import {} from 'next/navigation';
```
Comment (salmanR107) [inline app/components/home/home.js]: Remove

File: app/components/home/home.js [hunk 11]
```diff
@@ -12,19 +15,18 @@ export const HomePage = ({ initialPokemonDetails }) => {
   const [offset, setOffset] = useState(initialPokemonDetails.length);
   const [isLoading, setIsLoading] = useState(false);
 
-  const loadMore = async () => { 
+  // const router = useRouter();
```
Comment (salmanR107) [inline app/components/home/home.js]: Remove commented code

File: app/components/home/home.js [hunk 12]
```diff
@@ -37,11 +39,24 @@ export const HomePage = ({ initialPokemonDetails }) => {
...
+        loadMore={loadMore}
+        hasMore={hasMore}
+      >
         <div className={styles['pokemon-container']}>
           {pokemon.map((p) => (
             <div key={p.id} className={styles['pokemon-item']}>
-              <Image src={p.back_image} alt={p.name} width={150} height={150} />
+              <Link href={`/${p.name}`}>
+                <Image
+                  // onClick={() => router.push(`/${p.name}`)}
```
Comment (salmanR107) [inline app/components/home/home.js]: Remove this commented code if it is not needed

File: app/components/pokemon-details/pokemon-details.js [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+      <div>
+        <h2>Height</h2>
+        <p>{pokemonData.height}</p>
+      </div>
+      <div>
+        <h2>Forms</h2>
+      </div>
+      <div>
+        <h2>Moves</h2>
+        {pokemonData.moves.map((pokemon) => console.log(pokemon))}
```
Comment (salmanR107) [inline app/components/pokemon-details/pokemon-details.js]: Why are you logging it? Remove the code or display it

File: utils/fetch-data.js [hunk 14]
```diff
@@ -0,0 +1,18 @@
+export const fetchData = async ({ URL, query }) => {
```
Comment (salmanR107) [inline utils/fetch-data.js]: Please rename query to queryParams

**Review by salmanR107 — CHANGES_REQUESTED** [review]

File: app/[pokemonName]/page.jsx [hunk 15]
```diff
@@ -0,0 +1,15 @@
...
+import { fetchPokemonData } from '../../api/fetch-pokemon-data';
+import { notFound } from 'next/navigation';
+
+export const metadata = {
+  title: 'Pokemon Details',
+};
+
+export default async function Page({ params }) {
+  const data = await fetchPokemonData({ name: params.pokemonName });
+  console.log(data);
```
Comment (salmanR107) [inline app/[pokemonName]/page.jsx]: remove

**Review by salmanR107 — CHANGES_REQUESTED** [review]

File: api/fetch-pokemon-data.jsx (hunk 8 above)
Comment (zaini63033) [reply api/fetch-pokemon-data.jsx]: It's not getting the results if I remove the query parameter

**Review by salmanR107 — APPROVED** [review]

**Pull request merged**